feature/random statetest generator

tintinweb opened the pull request:
as per request @holiman 

added preliminary random statetest generator code to evmlab (not sure if its really a good fit here but lets see how this develops)

it currently is a bit overengineered (while missing the complex features) as I tried to stay close to the original cpp code. There's lots of things that can be simplified once we agree how to proceed.

not implmented:
* random_rlp - do we need random rlp code generator? using pyrlp may make this code os dependent.
* random_code - did not implement it yet as I am thinking about implementing random code generation to ethereum-dasm and using it from there.

cheers,
tintin

holiman commented:
Some context ... So what we want to produce are files like these: https://github.com/ethereum/tests/blob/develop/GeneralStateTests/stRandom2/randomStatetest386.json

The section `_info` I think we can just fill with whatever, `env` we can use the same thing every time, the `post` section I think we could fill like this:

```json
        "post" : {
            "Byzantium" : [
                {
                    "hash" : "0x00000000000000000000000000000000000000000000000000000000deadc0de",
                    "indexes" : {
                        "data" : 0,
                        "gas" : 0,
                        "value" : 0
                    },
                    "logs" : "0x00000000000000000000000000000000000000000000000000000000deadc0de"
                }
            ]
},
```
When a client runs this test, it will notice that "Oh my I'm failing it, because the post-state is wrong" -- but we don't care about that -- all we care about is that the trace and poststate for evm1 equals the trace and poststate for evm2. 

cc @guidovranken -- who is working on creating statetests out of libfuzzer corpus.

holiman commented:
Oh, 

* random_rlp -- not needed
* random_code -- needed. 

Other than that, we don't actually _need_ the templates for vm-tests or blockchain tests right now, but who knows in the future...

holiman commented:
Regarding random code generation. It could make sense to make it somewhat configurable, for example

- Make it configurable to that some certain opcodes are more prevalent (if we want to maximise coverage of new opcodes)
- Make it configurable so that it doesn't trip on invalid opcodes very often (but every once in a while those too)

holiman commented:
I replaced randomcode with a static `6060`, just to see how it looks. Tried the following: 
```
from evmlab.tools.statetests.randomtest import process_template
from evmlab.tools.statetests.templates import *
import json

print(json.dumps(process_template(TEMPLATE_STATETest)))
```
The json-dumping fails... I guess I'm doing it wrong -- how can I test this?

tintinweb commented:
Hi,

two options for dumping as json (the default json encoder does not know how to serialize our custom objects)

1. `json.dumps(template, default=str)`  - instructs the jsonencoder to apply str() on unknown objects
2. `json.dumps(template, cls=evmlab.tools.statetests.randomtest.RandomTestsJsonEncoder)` - instructs the jsonencoder to let `evmlab.tools.statetests.randomtest.RandomTestsJsonEncoder` handle unknown objects. this is atm quite easy with str but we might go for the class in the future to have more control over the encoding. That's why I added both examples.

tintinweb commented:
> Regarding random code generation. It could make sense to make it somewhat configurable, for example
> Make it configurable to that some certain opcodes are more prevalent (if we want to maximise coverage of new opcodes)
> Make it configurable so that it doesn't trip on invalid opcodes very often (but every once in a while those too)

@holiman definitely what I was thinking of. 
* training a neural network right now to get somewhat natural looking code with quirks. (just out of interest)
* thinking about adding a random code generator to my ethereum-dasm project, based on a distribution of source dumped from the chain. a bit more sophisticated that specifying weights. I'd also like the objectified instructions from that project to serialize valid code that does not break because of syntactic problems.

next steps:
* randomcodegenerator - as said, it'll likely be three different approaches that you can select in the end. starting with a completely random bytesequence generator, then to the weights and maybe the rnn thingie.

tintinweb commented:
https://github.com/ethereum/evmlab/pull/57/commits/fd5b6c698d75ea88c452319e9954f9a08a2b3a3d  implements the distribution based rndcode generator.

* ✔️  (inaccurate/likely invalid code) rndcode based on byte distribution (as implemeneted with https://github.com/ethereum/evmlab/pull/57/commits/fd5b6c698d75ea88c452319e9954f9a08a2b3a3d - this one will find issues with odd byte sequences
* TBD (more accurate/valid code) rndcode based on instructions (disassembling bytecode to instructions, count likelyhood of instructions, randomly serialize instructions based on the likelyhood) - this one should find problems with odd instruction sequences and random arguments to instructions
* TBD (more accurate/valid code) based on the trained RNN

holiman commented:
There is an existing constantinople-consensus bug in geth, for `SAR`. I'm really curious how long it'll take to find it, once we start fuzzing :)

holiman commented:
Some comments, based on the test-snippet above

```
...
    "transaction": {
      "data": [
        "'0x20005760616306576010017c60200000010060d601000060ee0004000400001400000000...
```
It double-encodes the data, with quote-chars and backticks. 

```
      "value": [
        "'0xf3ef243'",
```
Same double-quote problem, and also, I suspect it  _might_ be better to use `0x0f3ef243`, with zfilling so it's an even number of hexdigits. I believe geth won't accept odd hex input.  

Also, shouldn't that template which produces a 'filled' statetests be moved into the other templates (it seems internalized into a test-method?) ? That template is really the most important one. 

Great progress!

tintinweb commented:
* good point with the zfill, will look into that.
* the quoting issue is because I just hacked this to work for json.dumps. the dblquoting magic should only happen when pprinting. it's already on the list :)
* we'll move that template to the templates module once we agreed on the format there (template as object or template as string with placeholders)
* will be adding a cmdline interface later to make it easier to use.

tintinweb commented:
* added `RndCodeInstr()` that creates random code based on disassembled instruction distribution of evmcode taken from a `statedump.json`. 

we'll clean up the hacks later.
this requires a new version of ethereum_dasm to be installed. will push that version soon, just need to clean it up a bit first.

holiman commented:
> * once we agreed on the format there (template as object or template as string with placeholders) 

I would say let everything be objects as much as possible.

holiman commented:
Regarding the intelligence of random code -- there's no real need to be too intelligent, or alike real-world contracts. For example, normal contracts on chain seldom do SUICIDE, which are very VERY interesting during fuzz-testing, since a lot of corner cases / consensus issues have been caused by create/revert/suicide scenarios. 

If it does a `CREATE` or two, the resulting address should have a high likelihood of being called (which, I think it already does -- those addresses are included in the existing corpus, right?). 
If it does a `CALL`, it should have a high likelihood of calling into some code that exists, rather than a random nothing. 
Sometimes, but mostly not, we should have a too shallow stack to successfully call an operand. One consensus issue we found last time:    a shallow stack caused error (parity), while the same op on geth errored _earlier_ -- because the EVM was already in `static` mode of operation, so geth stopped it before the stack was checked. Since the op was `CREATE`, parity failed/exited the CREATE, whereas geth failed/exited the `STATICCALL` -- one call context up. 

So don't strive too much for imitating the live chain -- live chain transactions pretty controversial and boring :)

holiman commented:
regarding that sample, looks good:

```
[user@work statetests]$ ~/go/src/github.com/ethereum/go-ethereum/build/bin/evm --json statetest sample.json   
{"stateRoot": "418201ce3c47ccff7557aa00a8cb797e7bf39af2a1bc5ace38cb3a91b9af5db3"}
[
  {
    "name": "randomStatetest386",
    "pass": false,
    "fork": "Byzantium",
    "error": "post state logs hash mismatch: got 1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347, want 00000000000000000000000000000000000000000000000000000000deadc0de"
  }
]


```
(the reason I didn't get any json traces is that it calls `0x1000000000000000000000000000000000000000`, which is not defined in the `pre` section. But syntactically it seems fine!

tintinweb commented:
* HexInt values are now zfilled %2
* made ethereum_dasm optional until a new version is released and `RndCode` generator configurable (either the instruction based or bytedistribution based)
* moved templates to subpackage
* added CLI `python3 -m evmlab statetests --count=3`
* added seed tracking (basically dumping the internal randomizers state as there's no easy way to get a seed for a current state as its one-way). should allow to easily reproduce random states templates (as long as the general use of random stays the same ;))
* fixed quotes for json and pprint
* updated unittests

you can give it a try with (produces 3 json statetests - one per line. this is better performing than calling the process 3 times to generate 3 statetests)
`python3 -m evmlab statetests --count=3`

tintinweb commented:
> If it does a CREATE or two, the resulting address should have a high likelihood of being called (which, I think it already does -- those addresses are included in the existing corpus, right?).
If it does a CALL, it should have a high likelihood of calling into some code that exists, rather than a random nothing.
Sometimes, but mostly not, we should have a too shallow stack to successfully call an operand. One consensus issue we found last time: a shallow stack caused error (parity), while the same op on geth errored earlier -- because the EVM was already in static mode of operation, so geth stopped it before the stack was checked. Since the op was CREATE, parity failed/exited the CREATE, whereas geth failed/exited the STATICCALL -- one call context up.
I think we should probably have a call abt. that and align on what and how we want to tackle this.

> So don't strive too much for imitating the live chain -- live chain transactions pretty controversial and boring :) ((for example, I don't think live-chain even uses 'staticcall' or 'callcode' that much)
Just trying to keep the effort low (not having to specify weights by hand) and let stats and ai do the magic :p we can easily add more code generators and switch them in favor for others once the code is already usable.

holiman commented:
Nice! 
Regarding intelligence of code, we can iterate on that, and have different approaches/models. 

I tested it a bit, and wasn't able to get any traces -- I think the problem being the `to` address. Some sampling:
The `pre` accounts seem to always(?) be 
```
  "0x095e7baea6a6c7c4c2dfeb977efac326af552d87",
  "0x945304eb96065b2a98b57a48a06ae28d285a71b5",
  "0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b"
```
and the to-accounts seem to be 
```
      4 "0x1000000000000000000000000000000000000000"
      4 "0xb94f5374fce5edbc8e2a8697c15331677e6ebf0b"
      3 "0xc94f5374fce5edbc8e2a8697c15331677e6ebf0b"
      4 "0xd94f5374fce5edbc8e2a8697c15331677e6ebf0b"
      5 "0xffffffffffffffffffffffffffffffffffffffff"
```
Haven't looked into the cause, but there seems to be some mismatch in prestate versus what get's called. 
I would prefer that the `to` is either empty (contract creation) or one of the addresses that are either a precompile or exists in prestate.

tintinweb commented:
ah yeah, I was only using state accounts ([see](https://github.com/ethereum/evmlab/blob/823a6fd798794c5cda732941a75902a8c8b4e4cd/evmlab/tools/statetests/templates/object_based.py#L72)) as Destaddr ([defined here](https://github.com/ethereum/evmlab/blob/823a6fd798794c5cda732941a75902a8c8b4e4cd/evmlab/tools/statetests/rndval.py#L730))

tintinweb commented:
this is the current address distribution for `DESTADDRESS` (10000 samples) (implemented based on [the cpp reference](https://github.com/ethereum/testeth/blob/ee0c6776c01b09045a379220c7e490000dae9377/test/tools/fuzzTesting/fuzzHelper.cpp))

```
<RndAddressType.STATE_ACCOUNT: 4>        : 74.940000%
RANDOM                                   :  3.290000%
<RndAddressType.PRECOMPILED: 2>          :  1.410000%
<RndAddressType.SPECIAL_CREATE: 102>     : 15.860000%
<RndAddressType.BYZANTIUM_PRECOMPILED: 3>:  1.540000%
<RndAddressType.SENDING_ACCOUNT: 5>      :  2.960000%```

holiman commented:
Sweet, will do! Re nonce, though, I suspect address for any contracts created on the first (couple?) nonce is already in the corpus, so may be not much sense in using other nonces (or very seldom)

holiman commented:
Ah, I think I spotted the problem.. 
At https://github.com/ethereum/evmlab/pull/57/files#diff-002b49898c2448fffc2c1c22f0d1aa47R730 : 
```
                 RndAddressType.STATE_ACCOUNT: ["ffffffffffffffffffffffffffffffffffffffff",
                                                "1000000000000000000000000000000000000000",
                                                "b94f5374fce5edbc8e2a8697c15331677e6ebf0b",
                                                "c94f5374fce5edbc8e2a8697c15331677e6ebf0b",
                                                "d94f5374fce5edbc8e2a8697c15331677e6ebf0b"],
```
But in fact https://github.com/ethereum/evmlab/pull/57/files#diff-ad595afc13cf51c7898774b5689a55fdR41 : 
```
                    "0x095e7baea6a6c7c4c2dfeb977efac326af552d87" : { ...
                    },
                    "0x945304eb96065b2a98b57a48a06ae28d285a71b5" : {...
                    },
                    "0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b" : { ...
```
So the _actual_ prestate contains `095e`,`9453` and `a94f`, but the `STATE_ACCOUNTS` are defined as `b94f`,`c94f`,`d94f` as well as `1000` and `ffff`. 

The original template (https://github.com/ethereum/testeth/blob/develop/test/tools/fuzzTesting/createRandomTest.cpp#L241) contains prestates for all six of them.

holiman commented:
Copy-paste from discord about the original implementation:
So, I think it's here: https://github.com/ethereum/testeth/blob/7cbbb6fed4941420fbae738828fa1339c990e3d3/test/tools/fuzzTesting/fuzzHelper.cpp#L286 and the clause below, where it calls `fillArguments` , and if it's a [`CALL`](https://github.com/ethereum/testeth/blob/7cbbb6fed4941420fbae738828fa1339c990e3d3/test/tools/fuzzTesting/fuzzHelper.cpp#L391)
it makes sure the addr is `RandomCodeOptions::AddressType::PrecompiledOrState`
However, it only does this 'smart' filling if `rand < _options.smartCodeProbability` -- so sometimes it will be smart and fill ops so they 'make sense' (call existing accounts), other times it's just random.

tintinweb commented:
fixed the template structure and it now outputs transaction traces. 
good to merge.

will add smart code generation with another PR. 

cheers,
tin

https://colab.research.google.com/drive/1TILPcU5oFtfRJaMC8eBPZo8UU9shxr8I
```shell
{"randomStatetest0": {"_fuzz": {"compressed_random_state": null}, "env": {"currentCoinbase": "0xd649b47a4b279e5506e4b7d6b1054565255135dd", "currentDifficulty": "0x20000", "currentGasLimit": "0x32058a1f7e13f6", "currentNumber": "1", "currentTimestamp": "1000", "previousHash": "8685d4a020f5e24e39e86189b525bd184b9b09023280ab0928e6f3c93145f181"}, "post": {"Byzantium": [{"logs": "1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347"}]}, "pre": {"ffffffffffffffffffffffffffffffffffffffff": {"balance": "0xa29d06b9774dadb6", "code": "0x403553631560003500690041d200410060000052636061023200353c3ce6ee40000040701960f3000440006f0335002061d2195704006000018028007c57013cd26000606540596060190a606060001400000200a461003600000a5a6319378000c5a202340c00000000700061326136560060bd1506808100b160606020873690ff6960ff2aff60008015035656819561635010523000e039d72000004101f820520652b261010152030163e7ff208300ff17ff900f00605b506001903c105b476190ff501c60000051606081caff6000408457034003ee5451dbff5060c15161ff9020905484ee017fff505b02ff165480b6802081809001006003ff8061ff20800652150060ff0356815b51e1ffff60160292019b00200500a581528060739e9056a7ff5490ff61ff80ff00810004e0570090905b0401ffff495281030161600880200302ff2290e1d2500700028360ff5b6020000a83015180fcffff605400568580a200ff03003700164af160009250025b526082600804ff528100b7a101d00e004f3d49ff14047360619191ff52546054017f80ffff01836173c000010b6001ff50f660616104ff62508117ce506081035b9a0160004e90ffff010004803473ff3a94d95660600060500050ff606152025b10ff57d60016010005808160c15060506055905061ff609052010101578400a004000a7f86ff006110906050ffb65b60a18301ff0a00ff61a06190f168545451003901825b00ff60819201678240065b8850009114ff00ff16ffa033ff5449405b82507490606083ff7460ff048300ff019f50ffa080529032ff0115609a20905147905000ff6051004350026160ff54529163020100010020832000ff6171ff02ff00019050ff03ff8160616191808261900380000183659b010383005240020403548214ff6004610d0782608001900290817a0050605852951934ff61600016811651571d8116801f012300fff00290645b009161ff600019a08214ffb7ff807ab3900016c06061908060909050ff0016ff40ff27016c61ff8247039081ff915b815b1472015002006060ff19bb90ff600a60204200126185ff52375760ffff812060ff908168ff901f00618281ff405081002050ffe4046183ffa3606115546054900104016080600584605740408260619201175aa660905056005600ff0084209081b75780025060604061818156041616ce8261ff2380ff0200ff42ca6084500a0184501617a28190614402600401ff202050528351ff0a90ff0aff82936cff208457730120ff5055006060805b8160600a8061afba50a761ffa2a03160605793900151607a10575aff000550609790ff2a040401015b0301a9619102b060ff61015785ff612080ff31ff8341605260619100506100fc9d9080da6081606060006100200156606001500a409f543690616140516194610360600184810160e00061610002605615605b818294600a50ffff51600b918180094460610091601665044020927c321561cf545060f903ff029061080a50809050906060501901615b81603500ff600a0061812691068101c961606060ce60ff20ff2085a001ffff05ff509061636815ff60801501606160a0ffffff6001025101607182ff900552516002146060575a015b505249ca0f74038491cc80ff617f5420600182955702006385ff61a3ff60ff830024cc0216000150ff90ae5b635060ff94635a0f810460506115456079ff04ff02056084605b56809074506102916060a00360ff50501583fa618054ff60805b1650613a0183506001f1ffa60115ff64612015ff14039060ff616384610300ff810181611694508163158c208090151551575b0090b6550a52340181606eff01220446005001da57ff365b03ffd2614d565b5b6050405961ff3401169003879092566046a15b6054ff5601ff52611f5bff5057400a90505040ff506057046060018201f2916060602060505a82520003908060903bd20340f19100ef8680908060809020500001501e80825637346001141a5050", "nonce": "0x1c", "storage": {}}, "1000000000000000000000000000000000000000": {"balance": "0x5b02dee572d0078a", "code": "0x00157c4000605900d24789006000000057803800145264006060575770f200000000612300870000f3000061808036607bb0636057009057610000364015006045001eb5004100602200ea00fd148000f881000080005e6304e15a80a75a6015b59000365a00370140005a020000004860351560e40000000060f0810000000a6091c89061337a608111dd817300b3520160635b16600200002081f9838090e05b9002ff80200eff02ff009025207f040000511604905401609156816156016061c8604f50ff800161ff6157600901c58181beff95601990000283437f560460006015816001409180bb045002600180ff5b5660520060601520815b526060de556001de0482ff39a1002016000061500124508352606120566061ff5101e891155720f1596c830001030016ff903550546300506016018457013502842057566060245b076150409057aeffe00060011690605b6001017103506107020233ff056054016082ff00ff016f03900aff60ff0801ff1105916120129000606160516081ff605b60610c5260013700029401810d0317ff15ff605b51ff5044600060610390006016ff81ffff00600191205b2b00ff2450730052618383ff6e002439ff5e90021f111c3501aeff008b915635608583605081ca40509081ff800400ff19be73275b52320041050260569061059100186060000055ff8316835b80955761525b14040a00ff01006060a00a51946060925554ff5081161661db8100508081545a608055001092ff81c120ff031460046019ff60ffff93ff9081179060ff80a0600061b54811931990ff355101600f80610060541481400320910151ffffff5e4661c8732001008250600019ff206d003343008290605b045b602f029280ff9000e2600101c1816020305020508420000016016061e82082ff6039ff51608211170360606001ff804081ff82ff2000618060006082ff8081508182005469609285015280827d33918450819035400b5102801516805b60575b01505420006f0a109d02605ba0ff0133803540b65b616690635a60600014ffa620521101614290ff9080ff806006ff440461502004818300ff8400604e1050508060eba0a446509001200304600000902051000561158160606024601e50aa025e81c984600516601660849e9000520320600092e0506020ff73606da6056156510159610260ff1fffda3291036360610063556050608161ff7f160eea0276578120ff82806c5660604000ff91ff52ff01110000810151db90ff0240814390ffff9060913dff90ff01573544006057015bffff155180155705ff549090059154900083904256606bff01616061b5485b170091175061948157ff0000ff80935b025a044881160005cb5760050082615460ffeb0133570a0f01610190ff470057ff010457fc800157000287608a51608274021562823a1483ffff39003515608090c55480780082916180ff4d159082915b61ae820050ab2061ff546173803d56501914166011915181ff00610b811620ff008103605b8eff823e61ff0090ff5b8d205103bbd6acda00169f2037606132015b600150540160730211900a815b40015060168235735101a96048030280ff035b505b52ff60bd506091906040905b1f550aff8490818181519081ff815220ff93041f81735601010080902400515b7357a16163830851916084143557a820575b6001900049155001ffff016161606081c060519190ff0a32518015118160610063160017515b815160b4820090010182908b009c810452911c0160ff3360da045402540a32845083ff56ff011730550054048d5850ff0351541910ae5b8120566303606180ff0ba0608316195bf381045740509355908482ffff0a90614000671600819000799084ff6016822a420373206150800b50ff156090840203808251ff3700630281829bee1c04806111bb5b05ff05ff9001ff90906001826120ff6020910020f3516198523015522819605bc00160ff0106730143ff00000111a020ffff0101f29031350257590c9032029028203950ffeb60becf3581165cbc815250ff90915100ffff0186845b17fa15035b8705ccb4520160570490561560926191819000ff2083c2489dff165280015764ff010096808101805bd98016800173826060801602569261730187805251811050505b80a96090017f615690905261d6813d50500381ff615252606000010000336000ffff005b61ff5090901650c140509252608290918401600015900157505603614028618016600005818260ae6090ff5508606061556114615709026139559106ff195b006054560090805091351d0181c01650900055606160618102900350605550ff029090168290856a0450518150f65040616302632000515152011fffff226180609060156063202030ca60416d0120ff72ff630061165102a1015280615260022001a060a06090023683906060156150165b8efcff50353657ff83d6616060ff020001576001168560524d04200001140051808057c501b70a80ff02ff609081579402ff90606000615060a0600015e08215605681808301be005050ff5beb900060b85720e1ae61875002044f529015618200ff810181000a549060018201ff5be06101807360510301c80300806333150080920000008050f35081b690525460b42049f302545bff936060505b9160f1ff1960ff355260605050816073509081ff42602360504003605be481c0eecd00025660ff0080aa5059ffee5b5916ff3856ffa3ff005b5b572760606eff25358019028060569057600080325080808f9030405560ffb640ef00545a6190b75b605b919054ffb05b9293905a22ffe4406088916080d2616050022350502760f30139fc4734008156805003516061", "nonce": "0x1c", "storage": {}}, "b94f5374fce5edbc8e2a8697c15331677e6ebf0b": {"balance": "0x010025a2b7ff3a3d", "code": "0x6163006003d1d20057c0613557574002143787f20000405a605b00def200800a606061000100000060e4000060e036361314c70060c01117000001607c6000600004200300cd35e463030061be000024007c0000016020c100006060600000601460600a813100601480000572607c0060ee80c56324630000000000630160415b73ff850400558011505283608152929040605460600009ff0391a03910934c8820816003c9900a508460ff825bb081ca60019006635282ff828160605b825150635b361661ff18008144c8560090ff0020839060602bff20ff03666190ff528101920003010252009014800000b10191e361088814020100639451ff0055cafcad9c615c1f60a1f15b5749926000555057ffe15b2102ff8300600eff010551015100ff028357b00800eea79002018781500137ff8252205b022000560261ff00370ab60aff736081ff69032dff5635610640140eff54610194827f80815502819040ff87545081601403556023808101815081600181ffffff800060ff1400606081f317610556782c130057ff6030605463fd3020600182568020915681821452ffff408001f860160803e0156054ffff9017005056005b50910300000193605260613d5757005091045481815f0256157c500390b75260905b54568050ff8241526000a060009100845002607452529360b060c4be150680ff570057966b200302045180615bf86020ff6004ff570050560257501151c05408ff2010613701ff6e60f600005460816004600152616160000d2763a01873055060847e810050000a2081040a0001600160005b2000ff906017c660ffff22e8ff513401fe57506033165b540081dd900119143181089580ff00ff0000900303a2ff602001f3ff8710ff0050508b51d25506fff3030050900340010100616060005056351583ffff4002815177ff6060048695606015566020908182601083098160ff04cb17ff80010061a601501784605760615b427a0aff00810a020016526033033200a001502060616001600a3630ff0d520061905b02015b00511804ff370401ff825050901b521601c460006052606003619120732061ff3450e1ff160090ff81619050609083165152206092546120545201816016f361318060140060901600506161548101206004ef90555bff6003ff9088ff576014400100ff5160013281146060400081560250eb60520080c60316ff080194015082ff5b91908403020250fff3815050026115ff55ff3391140a0f3748140390202c0063ff90fc0061826052000100ef011257603740ff356174546040ff9100558190846016ff83ff02ff0016165b548060606103005052ff604c5b331b60805051520151ff5782206120605090005481606001906090936052090000906061150003ff82ff908402830351e504ff52115b20fff136f252715604505202500084006073528173920392933491605b0150160a540119170211579000ff0061728000605be0600156f06060ff5b00ffa70018cf63f360018103008163205793a6040301605290058f82600a60d66090a0000893010b90600140006181661416615008ff0f0360ffff330181fb01636068ff60ff0d52ff066060ff6060f1fbff50615c1704090061ffff01505181601401400188ff818000006090019001505003ff1500ff6161d301ff40109004d45091415b94529081815273326091827e00ffff20505061608150150081800061822061ff903ade009160acff9b635b5201a0a0345b206084602061606156153650605b8a000063000461ff119190510352ff6000ff511503ff903a00425081ee604082597390645063280af2ff50f200ff81ff01208288011551825b61ff81013c905690608160b060ff1680c0ff8856ff60555b51545b56320f905901609390c21661ff01606060200334d2ff00ff601f565a005a60f3f18181a9aa603e5beeff8251a250f2cb2b75ff60885090605b50015b5b605150", "nonce": "0x1c", "storage": {}}, "c94f5374fce5edbc8e2a8697c15331677e6ebf0b": {"balance": "0x20aec7385616c3aa", "code": "0x6102610061006087008b60486100950057616000610114000063600000004c60208161605715631460010000f33c0048600100000039600000ff0060006163140060205723610080606060736120000030000180906000600014367c81e1000080e16057600030e000cfa201870014049090a261600060904101800000008740510102208187168356600050015b556193ae500001b56064002200ff4000002003ae2080c72065601461816166006109045b599256546160806044014d6191600157854061ff9060035100519084936190556050206016154b6057ffe185000a8b601160835bff50606024911650015700f6036068800150506100ff90fd915202500400ff166016527d1640ff8d501656207a54606004606082618390500456165f601400fc5150ff56000b001591ff0015006060ff835201005a90700e0134900035ff611551609057800132ff5bdd948156605054da57916000525b60825641011660ff0761ff00548200ff57015b60911f037f5b910073fc785057575781f66056ff5250155c03568060ffffff911857fff23600ff0380ff15838260ff6040a04e03636152014757ffff835090544660169d0a160486605660146121ff6f00810360610092a202619056816d90161a20510080600201c73383ff6016576009ff9151ff5b816110606302c06020505017826091505063609001ff61dc6150ff00811f5161ff830160008050010201618d032451ff80ff00035100ff820001c93391b460566060856000918082508220010a0281603e50ff909090be50160f60021417545be811605a0a606160616001ff5090439055636080205b150087196f5290da6d619a9560ffff800136ff790001826116407f60601483ff24ff54bb812063518600245bd61cff81f45090376060816050190160ffa08290500085d2606054f1600160810103de812dc5f150ff6000358291576086ff6090008061ff15cb503c608060da33608091ff8088fffe83509190029081ff60226161116050828152816001008300002f9044ff1d515bff95085400926000860151158002618202525260501a60fd805401c6200215616001ffe7ff52ffff10b211ff6060ff9391ffeb0181615b81ff1391600200602060840a359081209252191556045090ff0affffd6928560c20b541d352e60ff0183b07f0a038056200081045b00928181008101500081ff820000508100686065b44902810054610102ff5acf140028ff810a60ff94015061805382ff81820000560a88608201000290605780014304900137606061603780ff520a7360529f8160ff6035046060f101c082601d52f20100026065906060800090105473560100828080ff019050568045603381150060368115610a005b60005b82ff010a0160829160928200ff00165001ff75ff00510191006091630000338120906060005400508160ff50827c5634f7010080ff60300160001587ff50805200ffff830119ff0160be526581005180602d005b011401ff56ff0356d29002605b01ff60030047906115909000046001ff605290015b1381008a5088800361019001915b2090008056ff820001f220000050604060248181810083ff81ff901500602090905757601f205123811563ff918164209251905250ff6063900052ff8390545b91166381609003409181ff020114812090400102600a5b81735026f55216167fff5b90600002936066ff80982000522016908101ffd604ff030060015116012040574c608180ff831461506000830260002a0061529833850101608081f15b0261200050ffff83899084f916820000015003e8129638351400208654406190030061518565618b76ff9161500a50ff8050ff500000ef16ff179181965654576003420c00541600ff33606082508283169500605500ff5b81515b208201ff1520ff6004ff614060ff650001005b600100009109370191549160406025810190006057d9205756565717ff60610445008100159020605b0960ffc850ff5414615180fd094e9051037390820202ff600d600006ffda60616000ffc1542b8315ff615252526060858280f8c954ff2a600390ff64609000005000905280545001145185092c03925400575580601080005bd2906020526028600a005257005060841660cd03009001730002015512200290016b154950ff9181606060560160611902842a006067504916906050800461030261905780526160ff47ffb48150021407ff5b0a830061505781ff14040940886000ff03171f60ff3261841650600181000360ff505254008160818100550000856002816052ff0c520a568000240090019040370016500056027396da608300506001505b00148119ffffff610a1452040003ff5b90606016360e903a0b0101fa0081012252609150d580817c05ff60bf90820001ff60e890ff57046150611656803602605b90966052400000908368011620ff01a39f8135606082ff0003ffff0a01830a096082ff000061027f604652daff836180ff0060010001605bee838150045bff6060a050005b608001240090814030908d4f60021501619182fa209000605060c25b36500082812050017861ff9080ff828201194990ff0380505080015660116082030101903690033720911120ff565b810190331fff505036200061509081610082935094513688903660ff0001ff90ee9140d40a6251000a91836190f30020559b005a6103d9365a", "nonce": "0x1c", "storage": {}}, "d94f5374fce5edbc8e2a8697c15331677e6ebf0b": {"balance": "0xfaf9fa6816f01fc7", "code": "0x375a35400000008100388b00b7608036e927346053616080b5360057220001a20080006156007e00196063608bb65700c00447616904008200005790686036006061148d006114010000416060ef00600060400a0200600080143660047f0063000a010361607a1460c2a2b500f86159e4d260001900a9603400007b15600060d6610451f2d280ff205b55605003cd60508182060481615b0080805b601f526000016000816050ff87376100526020505415039190010301515b1501ab83ffa05b508663ff018060020150ff04945261010d4060618090ff608d01201614ff4660608101605006745483992052001480009016ee315b16810a0900ff01014b0b605a5b90008415615b60809161330160601403019d81000a0a20169161f6841660ff905150ff19608820ee016100846040015beb5ba057638701d708ff92ff50153060ff4c508060837e01205056146004b814525001ff16209114203060eef380336005ff0190551691035290570080ff528100905bff20006057541d906001570b52206136db1563608250602061ff00545000500a6056048b516901825b576015ff600083801660ff60810002ff92606056608181608103e4405191019181ff81a001a66115600060609290915b8120523d165b0090018a0280200116fee59020fff1937161ff81ff8052400120b4ff5a01111f20928340910210ffeb005b005b0102009165016090ff020aff7601a0000060355481ff0035003f7a9080506154000251916091d852fd0080005450ef50618190ff50578180da501c8361206060e1010301506182960a5220905b704960915150ff700134ff80605161ff6081c6030182149050ff905191ff00334048105bff669080000060ff0060ff50900c0103819f50905b9050205b616009ff5460520160ff0001803b0190017f0174db00768052441501605700f1613653ff00552060026103165060002082a702801e42609081108092579015c10e44649190618456208280b420639251ff1482ffff5080030000a051ff00ff91898301225650ff836090605715609207ff5686001f906091838401ff6a605d4e56612020e18115f16b04da915616613960000157ff528e80355191ff0a57010014aa80fd154520ff61505260ff60ff86616060979020501f60ff0483f2036056619160909503c37b01fff4000090820600ffffaf0160618390503720fd0061816150ff9104021fffff9380009001565000848735816050018060600a61be908181505060822800b59060905b615451a0840852001601919154ffff6002ff018100ff03ff91ff0520ff5b57325b00826eff0060909002570f22608181908460018082005b5582ff10bd0101931f60505060000101909050827378805694ff605491ff6020ff61ff60540063c8526080205416bd0290818054ff8160ff005057015b5b155bf2650161040a81035193ff40910061509163ff20ff1f4d8260ffffaf80ff879003605bff80ae8051ff199f019114009081f28103039060ff028b0360611601915000528e8000ffffff90b35257612061ff548360ff3101019020ff005b90825700002000130901866050826152731560011680ff8160ffff906054005b61006009806181020156e792e18083905600ff0160575250610090336001609034608101ff910000328204ff6a61ff5200f0801db982500000608060ff005514ff506516923361816052608b3800604eff618681515ba96103e76050ffff509101826015005bff905b00605009814012ff60ff02820a0001730b505780835281f204ff60db506160104357895b610303606160ff5bff61400160f992635081400363800382ff81605201815b8298600181ff545200ff63816060600861005b610060610190ff9200d2708060909093829c925b5061ff39140060ff6055806080615082009046855604ff53ff5b50601152ff045b82ff6190d101035660e96102007c3661730a60168000d550829019906c605a015b5200508060576050811401da0250ff14408252507c40b40491006156ff606190000310836015600004288057ff5008201660ffbeff1d00160157ff615160520103806000502004ff90165b605201150f60e0810527606073630b604b575af30363ff8100fa5454814080016020831791906110810800110b8287605151ff0aae126091ff50519083800350010160000060600001540060d2c401ca81606d9055035001670000165bc2ff60140a0123ff5aff900150600057f2a700615b739000368782616050e50156ee04561f800107ff50cd1660602001674d36ff02ff94e21550ae305bff605660600056200050ff5661ff501450508260ff508202687e01d56020805c829056a0936060d7859061345066602250ff50", "nonce": "0x1c", "storage": {}}, "a94f5374fce5edbc8e2a8697c15331677e6ebf0b": {"balance": "0x25799541947f77d9", "code": "0x", "nonce": "0x1c", "storage": {}}}, "transaction": {"data": ["0x60fd3735006014220114151463e03700007c7c0000600057040060136104600040c0528a529000bd440000600035609500444490636000000060006e00610000009000363637000100000036ff007c270052180090600081520157606000000a403c0a5b27730000608000b500400215140a00603027047c00c219584100010affff91009080611660910061ff8061168081607a60507303248132ffff50011752ff52505402205090825b90ffff4101201900507d6020555790026c84200290e10400ffff57169f0c608290575b808020508481016090806150a01740405b612090914360005160800a504256401603613560019000ff228082910f60fd80866027900aff523660549080016000ffff3957186060000390ff008a81fd01905b508719ff6161810090610aa58b4060606350be035061010916ff5a60826016605051909002036081026a12816104809081ff0757528700148052ff610216ff00603582091f5220ffff1680ff61ff839250368181ff8052c863018151125bff000090009624ffff5b5516505955ff800f0150600100154c80500160105b90025035030161606020ff151600deb55754819100935b0183564881ff57566181908050ffff846080ff605bda5500ff5150cd600290820a81140161a0600a60e000e0407d57115bff83906000554000578460ffd581800000547f9091026051e4ff602001038000000f150260408056508360521f81026016505b0356ff15926e0001610061601550ff8018806056cb50ff60905f010f605b00ffa016ff014fff6060570054635080805683ff55411061526050916087ed025001ff92111650540411922060500020335060403133920101618161016000510161850180ca6003900a4050418261622060905b020360618200009283616016006056021f54406012000824925520f291ff010f80605603009152337f61815050036181ff605080d98480040160030860066060005060606601027360600502951550160632040f5720816003010102615190607e0040ff9001507360ff00209683ff0160026103ff16ff55b3921502a01420ff506003036081900300ffbb01c620b70016000c7c9147600150515c040bfc50dd5b085b03566335825200c060ffff81016002d00388808200528e37ff15618f90506102038091006002ffff5a35165600920050500a608300ff15ff40a08250c585508101604460ff140015900690161603159000566060400104578020609020ff005b8100ff80a8010d8201ee1490cc01409090155250ff01806160c35252206001200e02861673ff1681604020c00000577dd2ff01905059ff82016e41ff60008512ff8200000120ff17ff524950ff010160601015019034046d500201520150553756235023ff372060dd90005050608754609101ff50505660015080a82c20ffff157e6001ff34ff5b95b500ff1eff609080360160f15050019120519099ffff0300c2019087008160568190b0006136508217005050ff396004205056605636908081ff81f1d2ff61805657504f60b68150401f5b8d505656236160ff619036"], "gasLimit": ["0x9066a8", "3000000"], "gasPrice": "0x03", "nonce": "0x1c", "secretKey": "0x45a915e4d060149eb4365960e6a7a45f334393093061116b197e3240065ff2d8", "to": "0xd94f5374fce5edbc8e2a8697c15331677e6ebf0b", "value": ["0x9e83eed7", "0"]}}}
----execution----
{"pc":0,"op":55,"gas":"0x8f01b0","gasCost":"0x0","memory":"0x","memSize":0,"stack":[],"depth":1,"opName":"CALLDATACOPY","error":"stack underflow (0 \u003c=\u003e 3)"}
{"output":"","gasUsed":"0x8f01b0","time":147047,"error":"stack underflow (0 \u003c=\u003e 3)"}
{"stateRoot": "7e23e1e368d95ef59f68b85ac9d612625057793a35f454b14280731ec45074d7"}
[
  {
    "name": "randomStatetest0",
    "pass": false,
    "fork": "Byzantium",
    "error": "post state root mismatch: got 7e23e1e368d95ef59f68b85ac9d612625057793a35f454b14280731ec45074d7, want 0000000000000000000000000000000000000000000000000000000000000000"
  }
]
```
